File: trunk/src/Mobile/DotNet/PomaMobileApi/Connection/HttpSessionConnection.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Net;
using System.IO;

namespace Com.BkitMobile.Poma.Mobile.Api.Connection
{
    /// <summary>
    /// Represents a session connection through HTTP Protocol
    /// </summary>
    public class HttpSessionConnection
    {
        private WebProxy __WebProxy;
        /// <summary>
        /// Proxy of this connection
        /// </summary>
        public string WebProxy
        {
            get
            {
                if (__WebProxy == null) return "";
                return __WebProxy.Address.ToString();
            }
            set
            {
                __WebProxy = new WebProxy(value);
            }
        }
        /// <summary>
        /// Session ID of this connection
        /// </summary>
        public string SessionID
        {
            get;
            set;
        }

        /// <summary>
        /// Do a HTTP GET
        /// </summary>
        /// <param name="uri"></param>
        /// <returns>Result of <code>uri</code></returns>
        public string HttpGet(string uri)
        {
            // parameters: name1=value1&name2=value2	
            WebRequest webRequest = null;
            try
            {
                webRequest = WebRequest.Create(uri);
            }
            catch (Exception ex) { return ex.Message; }
            if (__WebProxy != null) webRequest.Proxy = __WebProxy;

            webRequest.Method = "GET";
            webRequest.Timeout = 30000;
            if (SessionID != null)
            {
                webRequest.Headers["Cookie"] = SessionID;
            }

            try
            { // get the response
                using (WebResponse webResponse = webRequest.GetResponse())
                {
                    if (webResponse == null)
                    { return null; }

                    string cookie = webResponse.Headers["Set-Cookie"];
                    if (cookie != null)
                    {
                        int semicolon = cookie.IndexOf(';');
                        SessionID = cookie.Substring(0, semicolon);
                    }

                    using (StreamReader sr = new StreamReader(webResponse.GetResponseStream()))
                    {
                        string s = sr.ReadToEnd().Trim();
                        sr.Close();
                        webResponse.Close();
                        return s;
                    }
                }
            }
            catch (WebException ex)
            {
                return ex.Message;
            }
        } // end HttpPost 

        /// <summary>
        /// Do a HTTP POST
        /// </summary>
        /// <param name="uri"></param>
        /// <returns>Result of <code>uri</code></returns>
        public string HttpPost(string uri, string parameters)
        {
            // parameters: name1=value1&name2=value2
            WebRequest webRequest = null;

            try
            {
                webRequest = WebRequest.Create(uri);
            }
            catch (Exception ex) { return ex.Message; }

            if (__WebProxy != null) webRequest.Proxy = __WebProxy;

            webRequest.ContentType = "application/x-www-form-urlencoded";
            webRequest.Method = "POST";
            webRequest.Timeout = 30000;

            if (SessionID != null)
            {
                webRequest.Headers["Cookie"] = SessionID;
            }
            byte[] bytes = Encoding.ASCII.GetBytes(parameters);
            Stream os = null;
            try
            { // send the Post
                webRequest.ContentLength = bytes.Length;   //Count bytes to send
                os = webRequest.GetRequestStream();
                os.Write(bytes, 0, bytes.Length);         //Send it
                os.Close();
            }
            catch (WebException ex)
            {
                return ex.Message;
            }
            finally
            {
                try { os.Close(); }
                catch { }
            }

            try
            { // get the response
                using (WebResponse webResponse = webRequest.GetResponse())
                {
                    if (webResponse == null)
                    { return null; }

                    string cookie = webResponse.Headers["Set-Cookie"];
                    if (cookie != null)
                    {
                        int semicolon = cookie.IndexOf(';');
                        SessionID = cookie.Substring(0, semicolon);
                    }

                    using (StreamReader sr = new StreamReader(webResponse.GetResponseStream()))
                    {
                        string s = sr.ReadToEnd().Trim();
                        sr.Close();
                        webResponse.Close();
                        return s;
                    }
                }
            }
            catch (WebException ex)
            {
                return ex.Message;
            }
        } // end HttpPost 
    }
}

File: src/Mobile/CLDC/src/com/bkitmobile/poma/mobile/api/connection/HttpSessionConnection.java
package com.bkitmobile.poma.mobile.api.connection;

import java.io.*;
import javax.microedition.io.*;

/**
 * Represents a session connection through HTTP Protocol
 * @author Hieu Rocker
 */
public class HttpSessionConnection {
	/**
	 * Session ID of this connection
	 */
	private String sessionID;

	/**
	 * Get SessionID of this connection
	 * @return
	 */
	public String getSessionID() {
		return sessionID;
	}
	
	/**
	 * Set SessionID of this connection
	 * @param sessionID
	 */
	public void setSessionID(String sessionID) {
		this.sessionID = sessionID;
	}

	/**
	 * Create an instance of HttpSessionConnection
	 */
	public HttpSessionConnection() {
	}
	
	/**
	 * Create an instance of HttpSessionConnection with specified session id
	 * @param sessionID
	 */
	public HttpSessionConnection(String sessionID) {
		this.sessionID = sessionID;
	}

	/**
	 * Request a HTTP GET
	 * @param url Url of Http GET
	 * @return Result of <code>url</code>
	 * @throws IOException
	 */
	public String httpGet(String url) throws IOException {
		HttpConnection httpConn = (HttpConnection) Connector.open(url);
		if (getSessionID() != null) {
			httpConn.setRequestProperty("Cookie", getSessionID());
		}
		
		httpConn.setRequestMethod(HttpConnection.GET);
		
		InputStream is = httpConn.openInputStream();
		String cookie = httpConn.getHeaderField("Set-Cookie");

		if (cookie != null) {
			int semicolon = cookie.indexOf(';');
			setSessionID(cookie.substring(0, semicolon));
			System.out.println(getSessionID());
		}

		int ch = 0;
		StringBuffer sb = new StringBuffer();
		while ((ch = is.read()) != -1) {
			sb.append((char) ch);
		}

		is.close();
		httpConn.close();
		return sb.toString();
	}

	/**
	 * Request a HTTP POST
	 * @param url Url of Http POST
	 * @param params parameters (param1=value1&param2=value2...) of the request
	 * @return Result of <code>url</code>
	 * @throws IOException
	 */
	public String httpPost(String url, String params) throws IOException {
		HttpConnection httpConn = (HttpConnection) Connector.open(url);
		if (getSessionID() != null) {
			httpConn.setRequestProperty("Cookie", getSessionID());
		}
		
		httpConn.setRequestMethod(HttpConnection.POST);
		httpConn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
		

		OutputStream os = httpConn.openOutputStream();
		os.write(params.getBytes());

		InputStream is = httpConn.openInputStream();
		String cookie = httpConn.getHeaderField("Set-Cookie");

		if (cookie != null) {
			int semicolon = cookie.indexOf(';');
			setSessionID(cookie.substring(0, semicolon));
			System.out.println(getSessionID());
		}

		int ch = 0;
		StringBuffer sb = new StringBuffer();
		while ((ch = is.read()) != -1) {
			sb.append((char) ch);
		}

		is.close();
		os.close();
		httpConn.close();
		return sb.toString();
	}
	
	/**
	 * Request a HTTP POST
	 * @param url Url of Http POST
	 * @param rawData 
	 * @return Result of <code>url</code>
	 * @throws IOException
	 */
	public String doPost(String url, byte[] rawData) throws IOException {
		HttpConnection httpConn = (HttpConnection) Connector.open(url);
		
		httpConn.setRequestMethod(HttpConnection.POST);
		httpConn.setRequestProperty("User-Agent", "POMA Mobile");		
//		httpConn.setRequestProperty("Content-Type", "application/octet-stream");
		
		OutputStream os = httpConn.openOutputStream();
		os.write(rawData);

		InputStream is = httpConn.openInputStream();

		int ch = 0;
		StringBuffer sb = new StringBuffer();
		while ((ch = is.read()) != -1) {
			sb.append((char) ch);
		}

		is.close();
		os.close();
		httpConn.close();
		return sb.toString();
	}
}